common/workunit/workunit.cpp

```diff
@@ -13641,6 +13645,11 @@ void executeThorGraph(const char * graphName, IConstWorkUnit &workunit, const IP
             }
             timelimit = runningTimeLimit;
             expectedStates = { WUStateWait };
+            unsigned __int64 blockedTime = elapsedTimer.elapsedNs();
+            {
+                Owned<IWorkUnit> w = &workunit.lock();
+                updateWorkunitStat(w, SSTgraph, graphName, StTimeBlocked, nullptr, blockedTime, wfid);
```

Comment:
I think you need to also cater for the : if (WUStateWait == state) // already finished
case above too.

Because it could in theory spend some time blocked, but then v. quickly transition from running to finished, so this code would see the state change to WUStateWait only, exit, and not record any blocked time.

Reply:
@jakesmith If it has already finished (i.e. it's WUStateWait==state) then we can't really say that it has blocked.  As it hasn't really blocked,  would we be misleadingly recording this as blocked time?

Reply:
@shamser - It is initially in WUStateBlocked state before it enters the for loop (line 13632).
It then expects after some blocked time, to transition to WUStateRunning.
When Thor finishes executing the graph, it will set the state to WUStateWait.

If the job is blocked for some time, but Thor executes the graph quickly, waitForWorkUnitToComplete will miss the transition from Running->Wait. It will only see Blocked->Wait.
The graph may have been blocked for an hour, but as it stands, it won't record any blocked time, because this code sees WUStateWait 1st and breaks out before it gets to the new code that records blocked time.

Reply:
Thanks for the explanation.  I'll make the change.